index.rst

```diff
@@ -13,17 +13,542 @@
+These could be implemented by storing quantum graphs and quantum status in files after workspaces are committed, while leaving room for database storage of provenance graphs as described in DMTN-205 :cite:`DMTN-205` in the future.
+
+Development mode
+----------------
+
+Workspaces would also be useful in pipeline software development and small-scale testing, especially if we could (sometimes) relax our consistency controls on software versions and pipeline configuration prior to "commit", a state we will call "developement mode".
+Furthermore, many development runs would not be committed at all, because the data products themselves are not the real outcome of that work - working code is.
+This suggests that workspaces should also support an analog of SQL's ``ROLLBACK`` as an alternative to commit (though we will use :meth:`WorkspaceButler.abandon` as the nomenclature here to avoid confusion with the very different relationship between ``commit`` and ``rollback`` in other contexts, e.g. some version control systems).
+
+External workspaces
+-------------------
+
+If we provide an option to create a workspace in an external location instead of inside a central data repository, we get an intriguing possibility for execution in the Rubin Science Platform (RSP) or on a science user's local compute resources.
+The quantum graph stored in the workspace knows about all needed input datasets, and these could be transferred to the external location as well, providing a self-contained proto-data-repository that could be manually moved and ultimately executed anywhere, and it contains all of the information needed to reconstitute a standalone data repository with its own database.
+Alternatively, if the workspace is left in a location where it can read from the central data repository but not write to it, input datasets could be left in the central data repository and the user could choose to either create a standalone repository that merely references them or commit by transferring output dataset file artifacts to the cental repository root.
+
+Sharded workspaces and Prompt Processing
+----------------------------------------
+
+We have one major use case that does not fit as well with this model, at least as described so far: prompt processing runs one ``visit`` or one ``{visit, detector}`` at a time, appending these results to a single :attr:`~lsst.daf.butler.CollectionType.RUN` collection that typically covers an entire night's processing.
+The switch from ``pipetask`` to workspaces wouldn't actually *require* any major changes to the prompt processing framework's architecture, because each prompt processing worker has its own full SQL-backed butler, with the fan-out to workers and long-term persistence represented as transfers between data repositories.
+Workspaces would be created and committed wholly within individual workers,  requiring some changes to low-level tooling like :class:`~lsst.ctrl.mpexec.SeparablePipelineExecutor`.
```

Comment:
We ought to be able to make it strictly faster, since prompt processing already has to write both QG files use per-worker SQLite DBs, and we'd be replacing the SQLite usage with more use of the QG and some much simpler and smaller JSON files.